services/brig/src/Brig/User/Auth.hs

```diff
@@ -71,19 +75,46 @@ lookupLoginCode phone = Data.lookupKey (userPhoneKey phone) >>= \case
 login :: Login -> CookieType -> ExceptT LoginError AppIO Access
 login (PasswordLogin li pw label) typ = do
     uid <- resolveLoginId li
+    checkRetryLimit uid
     Data.authenticate uid pw `catchE` \case
         AuthSuspended          -> throwE LoginSuspended
         AuthEphemeral          -> throwE LoginEphemeral
-        AuthInvalidCredentials -> throwE LoginFailed
-        AuthInvalidUser        -> throwE LoginFailed
+        AuthInvalidCredentials -> loginFailed uid
+        AuthInvalidUser        -> loginFailed uid
     newAccess uid typ label
 login (SmsLogin phone code label) typ = do
     uid <- resolveLoginId (LoginByPhone phone)
+    checkRetryLimit uid
```

Comment:
We could also remove this check, no? It's anyways checked when calling `loginFailed`, right?

Reply:
Not sure I understand, but let me try: if you only do the check in `loginFailed`, then after having exhausted your retry limit you can still successfully login.  You just need to eventually guess the password correctly.  So the check here is necessary.

Does that qualify as an answer?  Does it resolve anything?

Reply:
My point was: do we even need to call this `checkRetryLimit uid` here at all? In fact, is that function needed at all?

Calling `loginFailed` will call `decrRetryLimit` which will in turn reduce the budget in case of a failed login, right? And when that happens, we'll anyway throw `LoginBlocked`. Similarly as to how we do not explicitly check for the budget with SMS/Call

Reply:
summary of phone call: yes, both `checkRetryLimit` and `loginFailed` *in principle* might throw out-of-budget, but the latter is only ever called right after the former, so it won't.  the reason we need both is that `loginFailed` is only called in case of budget exhaustion or bad password, and we need to at least check for budget exhaustion in order for that to fail.  :-)

i welcome any ideas how to model this better!